### PR TITLE
cairo: Use freetype-sys instead of freetype.

### PR DESCRIPTION
**Pull request by songww**
#693

### CODE REVIEW

File: cairo/src/font/font_face.rs [hunk 1]
```diff
@@ -64,9 +64,7 @@ impl FontFace {
     /// `FontFace` is dropped.
     #[cfg(any(feature = "freetype", feature = "dox"))]
     #[doc(alias = "cairo_ft_font_face_create_for_ft_face")]
-    pub unsafe fn create_from_ft(
```
Comment (bilelmoussaoui) [inline cairo/src/font/font_face.rs:67]: I am not sure if it is a great idea to use the ffi bindings here

Comment (songww) [reply cairo/src/font/font_face.rs:67]: I'm not sure that changing to a safe method is an appropriate change, it's a big change.

**Comment by sdroege** [comment]
Using the safe freetype-rs instead of freetype-sys is probably nicer, or not?

File: cairo/src/font/font_face.rs [hunk 2]
```diff
@@ -87,15 +94,19 @@ impl FontFace {
...
-        ));
-        let status = ffi::cairo_font_face_status(font_face.to_raw_none());
+        let mut face = face.clone();
+        let font_face = unsafe {
+            FontFace::from_raw_full(ffi::cairo_ft_font_face_create_for_ft_face(
+                face.raw_mut() as freetype::ffi::FT_Face as *mut _,
+                load_flags,
+            ))
+        };
+        font_face.set_user_data(&FT_FACE_KEY, Rc::new(face))?;
```
Comment (sdroege) [inline cairo/src/font/font_face.rs:108]: Why is the `Rc` needed here?

**Review by sdroege — COMMENTED** [review]
Looks good to me generally

**Comment by jf2048** [comment]
this is only running 11 checks right now, please re-run the CI on this after https://github.com/gtk-rs/gtk-rs-core/pull/695

File: cairo/src/font/font_face.rs [hunk 3]
```diff
@@ -64,14 +69,16 @@ impl FontFace {
     /// `FontFace` is dropped.
     #[cfg(any(feature = "freetype", feature = "dox"))]
     #[doc(alias = "cairo_ft_font_face_create_for_ft_face")]
-    pub unsafe fn create_from_ft(
```
Comment (RealKC) [inline cairo/src/font/font_face.rs:67]: If you're turning this into a safe method, you should remove the `# Safety` section from the docs

File: cairo/src/font/font_face.rs [hunk 4]
```diff
@@ -64,14 +69,16 @@ impl FontFace {
...
-        face: freetype_crate::freetype::FT_Face,
-    ) -> Result<FontFace, Error> {
-        let font_face = FontFace::from_raw_full(ffi::cairo_ft_font_face_create_for_ft_face(
-            face as *mut _,
-            0,
-        ));
-        let status = ffi::cairo_font_face_status(font_face.to_raw_none());
+    pub fn create_from_ft(face: &freetype::face::Face) -> Result<FontFace, Error> {
+        let mut face = face.clone();
+        let font_face = unsafe {
```
Comment (RealKC) [inline cairo/src/font/font_face.rs:74]: Can you add a safety comment here? I missed the `.clone()` call at first and was worried about potential use-after-frees

**Review by RealKC — CHANGES_REQUESTED** [review]

File: cairo/src/font/font_face.rs (hunk 4 above)
Comment (sdroege) [reply cairo/src/font/font_face.rs:73]: Better would be to use `// SAFETY: blablabla` but this also works :)

File: cairo/src/font/font_face.rs [hunk 5]
```diff
@@ -87,15 +94,19 @@ impl FontFace {
...
+        face: &freetype::face::Face,
         load_flags: libc::c_int,
     ) -> Result<FontFace, Error> {
-        let font_face = FontFace::from_raw_full(ffi::cairo_ft_font_face_create_for_ft_face(
-            face as *mut _,
-            load_flags,
-        ));
-        let status = ffi::cairo_font_face_status(font_face.to_raw_none());
+        let mut face = face.clone();
+        let font_face = unsafe {
```
[the remark this answers is not in the document]
Comment (sdroege) [reply cairo/src/font/font_face.rs:100]: Same here about the format, should ideally be `// SAFETY: blablabla`

File: cairo/src/font/font_face.rs (hunk 2 above)
Comment (sdroege) [reply cairo/src/font/font_face.rs:106]: Indeed, thanks :)

**Review by sdroege — APPROVED** [review]
Looks good to me apart from the comments. Want to fix those @songww or otherwise I'll just do it before merging.

**Comment by sdroege** [comment]
Uh, new beta with new clippy warnings happened. Let's worry about those separately.

**Comment by sdroege** [comment]
Thanks for the improvements :)

**Comment by songww** [comment]
Thank you all for your advice！

**Pull request merged**